cmd/ursrv/privacy.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (C) 2021 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"fmt"
+	"math"
+)
+
+var (
+	privMin = 10
+)
+
+// Helper functions for private numbers and percentages
+func isPrivate(n int) bool {
+	return n < privMin
+}
+
+func privCount(x int) int {
+	if isPrivate(x) {
+		x = privMin
+	}
+	return x
+}
+
+func privCountString(n int) string {
+	if isPrivate(n) {
+		return fmt.Sprintf("< %d", privCount(n))
+	}
+	return fmt.Sprintf("%d", n)
+}
+
+func privCounts(x map[string]int) map[string]intString {
+	fGroup := make(map[string]intString)
+	for key, counts := range x {
+		fGroup[key] = intString{
+			privCount(counts),
+			privCountString(counts),
+		}
+	}
+	return fGroup
+}
+
+func privPct(x, total int) float64 {
+	return (100 * float64(privCount(x))) / float64(total)
+}
+
+func privPctPct(x float64, private bool) float64 {
+	if private {
+		x *= 1.01
+	}
+	return x
+}
```

Comment:
In some cases we need a private percentage, but don't have the total or count, only the computed percentage. This is an attempt at perturbing the percentage in order to preserve privacy. I realize now that this should be multiplication by a random value e.g. between 1.01 and 1.05

I couldn't think of a better name at the time and decided to use a stupid name and get on with the coding. What function name would you suggest?

Reply:
I don't think we need to add anything to small percentages where we don't have a count (where does the percentage come from?). What is the privacy implication of saying that 0.01 percent of an unknown number of something is whatever?

Reply:
This is called in the `group` function in `analyticts.go`, for "versions", "platforms", and "compilers". Knowing the percentage and total group size, one can infer the group size of any percentage. One way to prevent an adversary from deducing the group size is by pertubation.